content/blog/2022-05-05-e2e-cv-pt3.md

```diff
@@ -0,0 +1,292 @@
+      ```
+
+      This will return some http-header info and the body of the response
+      containing the defect segmentation mask (`0` for pixel locations without
+      defects and `1` otherwise):
+
+      ```json
+      *   Trying 18.208.60.216:443...
+      * Connected to demo-api-mag-tiles-dev.herokuapp.com (18.208.60.216) port 443 (#0)
+      ...
+      {"pred":[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+      0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+      0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+      1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+      ...
+      ```
+
+      Alternatively, we can do a similar thing with a Python script that also
+      saves the output mask into a `exp3_num_258558_mask.png` image:
+
+      ```python
+      import json
+      from pathlib import Path
+
+      import matplotlib.cm as cm
+      import matplotlib.pyplot as plt
+      import numpy as np
+      import requests
+
+      url = 'https://demo-api-mag-tiles-dev.herokuapp.com/analyze'
+      file_path = Path(
+          'data/MAGNETIC_TILE_SURFACE_DEFECTS/test_images/exp3_num_258558.jpg')
+      files = {'image': (str(file_path), open(file_path, 'rb'), "image/jpeg")}
+      response = requests.post(url, files=files)
+      data = json.loads(response.content)
+      pred = np.array(data['pred'])
+      plt.imsave(f'{file_path.stem}_mask.png', pred, cmap=cm.gray)
+      ```
+
+      Below you can see what this mask looks like.
+      ![Output mask `exp3_num_258558_mask.png`](/uploads/images/2022-05-05/exp3_num_258558_mask.png '=300')
+
+3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/3-deploy.yaml)
+
+   ![Workflow for deploying to the production environment](/uploads/images/2022-05-05/workflow_prod.png '=800')
+
+   Before we merge the dev branch into the main branch, we would need to
+   thoroughly test and monitor the application in the development environment. A
+   good test could be duplicating real image requests to the dev endpoint (here
+   it's
+   [https://demo-api-mag-tiles-**dev**.herokuapp.com/analyze](https://demo-api-mag-tiles-dev.herokuapp.com/analyze))
+   for some period of time and assess the quality of the returned segmentation
+   masks.
+
+   If there are no issues and we are confident in the quality of the new model,
+   we can merge the development branch into the main branch of our repository.
+   Again, this triggers the third CI/CD workflow that deploys the code from the
+   main branch to the production API. This looks identical to the deployment
+   into the development environment, except now the deployment endpoint will be
+   [https://demo-api-mag-tiles-**prod**.herokuapp.com/analyze](https://demo-api-mag-tiles-prod.herokuapp.com/analyze).
```

Comment:
Oh.  I see you are pointing out the difference in the links.  Maybe it would be better as:
"A good test... dev endpoint for some period of time and assess the quality of the returned segmentation masks.  Here it's https://demo-api-mag-tiles-dev.herokuapp.com/analyze (note the dev distinction).

And then similar treatment for the second?